src/main/java/gregtech/common/covers/filter/FluidFilterContainer.java

```diff
@@ -1,128 +1,175 @@
+            @Override
+            public ModularPanel createUI(ModularPanel mainPanel, GuiSyncManager syncManager) {
+                getFilter().setMaxTransferSize(getMaxTransferSize());
+                return getFilter().createPopupPanel(syncManager);
+            }
+        };
+        manager.syncValue("filter_panel", panel);
+        var filterButton = new ButtonWidget<>();
+        filterButton.setEnabled(hasFilter());
+
+        return new Row().coverChildrenHeight()
+                .marginBottom(2).widthRel(1f)
+                .child(new ItemSlot()
+                        .slot(SyncHandlers.itemSlot(filterInventory, 0)
+                                .filter(FilterTypeRegistry::isFluidFilter)
+                                .changeListener((newItem, onlyAmountChanged, client, init) -> {
+                                    if (newItem.isEmpty() || FilterTypeRegistry.isFluidFilter(newItem)) {
+                                        onFilterSlotChange(true);
+                                    }
+                                })
+                                .singletonSlotGroup(101))
+                        .onUpdateListener(w -> {
+                            if (!hasFilter() && panel.isPanelOpen()) {
+                                panel.closePanel();
+                            }
+                        }, true)
+                        .size(18).marginRight(2)
+                        .background(GTGuiTextures.SLOT, GTGuiTextures.FILTER_SLOT_OVERLAY))
+                .child(filterButton
+                        .setEnabledIf(w -> hasFilter())
+                        .onMousePressed(i -> {
+                            boolean success = false;
+                            if (!panel.isPanelOpen()) {
+                                panel.openPanel();
+                                success = true;
+                            } else if (panel.isValid()) {
+                                panel.closePanel();
+                                success = true;
+                            }
+                            Interactable.playButtonClickSound();
+                            return success;
+                        }))
+                .child(IKey.dynamic(() -> hasFilter() ?
+                        getFilterInventory().getStackInSlot(0).getDisplayName() :
+                        IKey.lang("metaitem.fluid_filter.name").get())
+                        .alignment(Alignment.CenterRight).asWidget()
+                        .left(36).right(0).height(18));
+    }
 
-        this.filterWrapper.initUI(y + 15, widgetGroup);
-        this.filterWrapper.blacklistUI(y + 15, widgetGroup, () -> true);
+    @Override
+    public void readInitialSyncData(@NotNull PacketBuffer packetBuffer) {
+        var stack = ItemStack.EMPTY;
+        try {
+            stack = packetBuffer.readItemStack();
+        } catch (IOException ignore) {}
+        this.filterInventory.setStackInSlot(0, stack);
```

Comment:
I think this logic should be in the super class, and an method providing access to the stack should be called for implementations to handle instead of having an unwritten assumption that all implementations of `readInitialSyncData` will read the stack.